libvast/src/system/connector.cpp

```diff
@@ -0,0 +1,178 @@
+    case caf::sec::cannot_publish_invalid_actor:
+    case caf::sec::cannot_spawn_actor_from_arguments:
+    case caf::sec::end_of_stream:
+    case caf::sec::no_context:
+    case caf::sec::unknown_type:
+    case caf::sec::no_proxy_registry:
+    case caf::sec::runtime_error:
+    case caf::sec::remote_linking_failed:
+    case caf::sec::cannot_add_upstream:
+    case caf::sec::upstream_already_exists:
+    case caf::sec::invalid_upstream:
+    case caf::sec::cannot_add_downstream:
+    case caf::sec::downstream_already_exists:
+    case caf::sec::invalid_downstream:
+    case caf::sec::no_downstream_stages_defined:
+    case caf::sec::stream_init_failed:
+    case caf::sec::invalid_stream_state:
+    case caf::sec::unhandled_stream_error:
+    case caf::sec::bad_function_call:
+    case caf::sec::feature_disabled:
+    case caf::sec::cannot_open_file:
+    case caf::sec::socket_invalid:
+    case caf::sec::socket_disconnected:
+    case caf::sec::socket_operation_failed:
+    case caf::sec::unavailable_or_would_block:
+    case caf::sec::malformed_basp_message:
+    case caf::sec::serializing_basp_payload_failed:
+    case caf::sec::redundant_connection:
+    case caf::sec::remote_lookup_failed:
+    case caf::sec::no_tracing_context:
+    case caf::sec::all_requests_failed:
+    case caf::sec::field_invariant_check_failed:
+    case caf::sec::field_value_synchronization_failed:
+    case caf::sec::invalid_field_type:
+    case caf::sec::unsafe_type:
+    case caf::sec::save_callback_failed:
+    case caf::sec::load_callback_failed:
+    case caf::sec::conversion_failed:
+    case caf::sec::connection_closed:
+    case caf::sec::type_clash:
+    case caf::sec::unsupported_operation:
+    case caf::sec::no_such_key:
+    case caf::sec::broken_promise:
+    case caf::sec::connection_timeout:
+    case caf::sec::action_reschedule_failed:
+      return true;
+    case caf::sec::incompatible_versions:
+    case caf::sec::incompatible_application_ids:
+      return false;
+  }
+  return false;
+}
+
+bool is_recoverable_error(const caf::error& err) {
+  if (err.category() != caf::type_id_v<caf::sec>)
+    return true;
+  const auto err_code = std::underlying_type_t<caf::sec>{err.code()};
+  auto err_enum = caf::sec{caf::sec::none};
+  if (!caf::from_integer(err_code, err_enum)) {
+    VAST_WARN("unable to retrieve error code for a remote node connection "
+              "error:{}",
+              err);
+    return true;
+  }
+  return is_recoverable_error_enum(err_enum);
+}
+
+bool should_retry(const caf::error& err,
+                  std::size_t remaining_connection_attempts) {
+  return remaining_connection_attempts > 0 && is_recoverable_error(err);
+}
+
+} // namespace
+
+connector_actor::behavior_type
+connector(connector_actor::pointer self, std::uint64_t max_connection_attempts,
+          caf::timespan connection_retry_delay, vast::port port,
+          std::string host, caf::timespan connection_timeout) {
```

Comment:
Can we wrap these 5 arguments into a struct and pass them to the `atom::connect` handler instead? I don't think any of them are needed at spawn time of the actor, and if you pass the handler then the actor actually supports connecting to multiple different endpoints out of the box.